service/storage_service.cc

```diff
@@ -6478,6 +6517,28 @@ future<join_node_request_result> storage_service::join_node_request_handler(join
             } catch (wait_for_ip_timeout& ex) {
                 rtlogger.warn("Failed to check liveness for replaced id {}. Asumming it is dead, Error: {}", *params.replaced_id, ex);
             }
+
+            auto replaced_it = _topology_state_machine._topology.normal_nodes.find(*params.replaced_id);
+            if (replaced_it == _topology_state_machine._topology.normal_nodes.end()) {
+                result.result = join_node_request_result::rejected{
+                    .reason = ::format("Cannot replace node {} because it is not in the 'normal' state", *params.replaced_id),
+                };
+                co_return result;
+            }
```

Comment:
We only pace a request into the queue here, so theoretically you can have anode that is still joining and then you can run a replace on it which will start after the node completes joining (it needs to die after that though for replace to succeed).

Reply:
Is it a problem? We already check that a node being replaced is dead here, and theoretically it could die later but before the topology coordinator handles the request.

I had to do something if a node being replaced is not a normal node. Otherwise, I wouldn't have been able to add checks that prohibit replacing a zero-token node with a token-owning node and vice versa.

Ideally, these checks could be done on the topology coordinator side (in `validate_joining_node`). Then, there would be no problem. However, the topology coordinator doesn't know if a replacing node is zero-token because `replace_param` doesn't contain tokens. Even if we add the `zero_token` boolean to `system.topology` as we discussed on the daily meeting, it won't help here unless we add this boolean to the in-memory state machine.

Reply:
> Is it a problem? 

Not a problem but an observation. IIRC we moved most of the checks to the coordinator to make it close to when a request actually handled. May be for replace and removenode it is find to do it while a request is posted since those should be done on a dead node anyway. But we need to have checks in the topology coordinator as well to handle cases where the same node is replaced/removed twice for instance. I do not think we check it when we place a request.

Reply:
If we replace/remove a node for the second time, the node has already been removed from normal nodes. We check that a replaced node is normal in `validate_joining_node`.

However, it looks like we don't handle the removenode case. I checked that if we concurrently initiate two removenode operations of the same node twice, one of them will hang. We check that the node being removed is normal in `storage_service::raft_removenode`, but we should also do it on the coordinator side. The question is how the topology coordinator should behave. I think it should mark the request as failed (the `rtbuilder.done("some error message")` thing).

I'm not sure if decommission and rebuild have similar problems, since the transition node initiates the operation by itself. Anyway, I think we can have a single validation for all these operations in `handle_node_transition`, `case node_state::normal:`.

I can open an issue to track this problem.

Reply:
> If we replace/remove a node for the second time, the node has already been removed from normal nodes. We check that a replaced node is normal in `validate_joining_node`.
> 
> However, it looks like we don't handle the removenode case. I checked that if we concurrently initiate two removenode operations of the same node twice, one of them will hang. We check that the node being removed is normal in `storage_service::raft_removenode`, but we should also do it on the coordinator side.

Exactly. So we forgot this case.

> The question is how the topology coordinator should behave. I think it should mark the request as failed (the `rtbuilder.done("some error message")` thing).

Yes.

> 
> I'm not sure if decommission and rebuild have similar problems, since the transition node initiates the operation by itself. Anyway, I think we can have a single validation for all these operations in `handle_node_transition`, `case node_state::normal:`.

You cannot start to decommissions on the same node twice.

> 
> I can open an issue to track this problem.

And send the patch? :)

Reply:
> > I can open an issue to track this problem.
> 
> And send the patch? :)

I could, but it looks like a good first issue (to read the topology coordinator code, etc.), so @kbr-scylla could have a different idea

Reply:
I opened #20271